rules/prefer-global-this.js

```diff
@@ -0,0 +1,259 @@
+	'personalbar',
+	'scrollbars',
+	'statusbar',
+	'toolbar',
+	'status',
+	'close',
+	'closed',
+	'stop',
+	'focus',
+	'blur',
+	'frames',
+	'length',
+	'top',
+	'opener',
+	'parent',
+	'frameElement',
+	'open',
+	'originAgentCluster',
+	'postMessage',
+
+	// Events commonly associated with "window"
+	...[...windowSpecificEvents].map(event => `on${event}`),
+
+	// To add/remove/dispatch events that are commonly associated with "window"
+	// https://www.w3.org/TR/DOM-Level-2-Events/events.html#Events-flow
+	'addEventListener',
+	'removeEventListener',
+	'dispatchEvent',
+
+	// https://dom.spec.whatwg.org/#idl-index
+	'event', // Deprecated and quirky, best left untouched
+
+	// https://drafts.csswg.org/cssom-view/#idl-index
+	'screen',
+	'visualViewport',
+	'moveTo',
+	'moveBy',
+	'resizeTo',
+	'resizeBy',
+	'innerWidth',
+	'innerHeight',
+	'scrollX',
+	'pageXOffset',
+	'scrollY',
+	'pageYOffset',
+	'scroll',
+	'scrollTo',
+	'scrollBy',
+	'screenX',
+	'screenLeft',
+	'screenY',
+	'screenTop',
+	'screenWidth',
+	'screenHeight',
+	'devicePixelRatio',
+]);
+
+const webWorkerSpecificAPIs = new Set([
+	// https://html.spec.whatwg.org/multipage/workers.html#the-workerglobalscope-common-interface
+	'addEventListener',
+	'removeEventListener',
+	'dispatchEvent',
+
+	'self',
+	'location',
+	'navigator',
+	'onerror',
+	'onlanguagechange',
+	'onoffline',
+	'ononline',
+	'onrejectionhandled',
+	'onunhandledrejection',
+
+	// https://html.spec.whatwg.org/multipage/workers.html#dedicated-workers-and-the-dedicatedworkerglobalscope-interface
+	'name',
+	'postMessage',
+	'onconnect',
+]);
+
+/**
+Report the node with a message.
+
+@param {import('eslint').Rule.RuleContext} context
+@param {import('estree').Node} node
+@param {string} value
+*/
+function report(context, node, value) {
+	const fix = (/** @type {import('eslint').Rule.RuleFixer} fixer */ fixer) => fixer.replaceText(node, 'globalThis');
+	context.report({
+		node,
+		messageId: MESSAGE_ID_ERROR,
+		data: {value, replacement: 'globalThis'},
+		fix,
+		suggest: [{messageId: MESSAGE_ID_SUGGESTION, data: {value, replacement: 'globalThis'}, fix}],
+	});
+}
+
+/**
+Handle nodes and check if they should be reported.
+
+@param {import('eslint').Rule.RuleContext} context
+@param {import('estree').Node | Array<import('estree').Node>} nodes
+*/
+function handleNodes(context, nodes) {
+	if (!Array.isArray(nodes)) {
+		nodes = [nodes];
+	}
+
+	for (const node of nodes) {
+		if (node && node.type === 'Identifier' && globalIdentifier.has(node.name)) {
+			const variable = findVariableInScope(context.sourceCode.getScope(node), node.name);
+			if (!variable) {
+				report(context, node, node.name);
+			}
+		}
+	}
+}
+
+/**
+Check if the node is a window-specific API.
+
+@param {import('estree').MemberExpression} node
+@returns {boolean}
+*/
+const isWindowSpecificAPI = node => {
+	if (node.type !== 'MemberExpression') {
+		return false;
+	}
+
+	if (node.object.name !== 'window' || node.property.type !== 'Identifier') {
```

Comment:
Need add `!node.computed` here, don't forgot to test.

Reply:
I think `!node.computed` should not be added here, or am I missing some case?.

Regardless of whether it is a computed property or not, this rule should be applied.

There is the test that we already have

```js
window['title'] // node.computed === true
window.title    // node.computed === false
```

```md
## invalid(4): window.foo

> Input

    `␊
      1 | window.foo␊
    `

> Output

    `␊
      1 | globalThis.foo␊
    `

> Error 1/1

    `␊
    > 1 | window.foo␊
        | ^^^^^^ Prefer \`globalThis\` over \`window\`.␊
    `

## invalid(5): window[foo]

> Input

    `␊
      1 | window[foo]␊
    `

> Output

    `␊
      1 | globalThis[foo]␊
    `

> Error 1/1

    `␊
    > 1 | window[foo]␊
        | ^^^^^^ Prefer \`globalThis\` over \`window\`.␊
    `

## invalid(6): window["foo"]

> Input

    `␊
      1 | window["foo"]␊
    `

> Output

    `␊
      1 | globalThis["foo"]␊
    `

> Error 1/1

    `␊
    > 1 | window["foo"]␊
        | ^^^^^^ Prefer \`globalThis\` over \`window\`.␊
    `
```

Reply:
`window[title]`

Reply:
There is no difference between `window[title]` and `window[foo]`

Both `title` and `foo` are variable accesses. test added.

```md
## invalid(6): window[title]

> Input

    `␊
      1 | window[title]␊
    `

> Output

    `␊
      1 | globalThis[title]␊
    `

> Error 1/1

    `␊
    > 1 | window[title]␊
        | ^^^^^^ Prefer \`globalThis\` over \`window\`.␊
```

Reply:
`window[name]` doesn't mean it's accessing `window.name`, since `name` can be other values.

Reply:
We are not checking `window` is `MemberExpression.object` either, `foo[window]`

Reply:
Please don't click `resolve` button, I can't find my comment.

Reply:
> We are not checking `window` is `MemberExpression.object` either, `foo[window]`

The test added in 4be46990636a294b5fa45064a478a4ecbdddeda8